crates/vm/levm/src/db/gen_db.rs

```diff
@@ -189,6 +170,27 @@ impl<'a> VM<'a> {
         Ok(())
     }
 
+    /// Gets original storage value of an account, caching it if not already cached.
+    /// Also saves the original value for future gas calculations.
+    pub fn get_original_storage(&mut self, address: Address, key: H256) -> Result<U256, VMError> {
+        let value_pre_tx = match self.storage_original_values.get(&address).cloned() {
+            Some(account_storage) => match account_storage.get(&key) {
+                Some(value) => *value,
+                None => self.get_storage_slot(address, key)?,
+            },
+            None => self.get_storage_slot(address, key)?,
+        };
+
+        // Add it to the original values if it wasn't already there
+        self.storage_original_values
+            .entry(address)
+            .or_default()
+            .entry(key)
+            .or_insert(value_pre_tx);
+
+        Ok(value_pre_tx)
```

Comment:
If we found the key, we can return early from the function

```suggestion
        if let Some(account_storage) = self.storage_original_values.get(&address) {
            if let Some(value) = account_storage.get(&key) {
                return Ok(*value);
            }
        }

        let value = self.get_storage_slot(address, key)?;

        self.storage_original_values
            .entry(address)
            .or_default()
            .insert(key, value);

        Ok(value)
```

Reply:
Changed it [here](https://github.com/lambdaclass/ethrex/pull/2629/commits/18b077533f17c1ceeccb646639659cb7861880c2).
It is slightly different but the idea of the early return remains.